## Script Contents:

Part 1: Exercise

Part 2: Task

1. Use df.describe to check for consistency in df_prods
2. Check for mixed-type data in df_ords
3. Run a check for missing values on df_ords dataframe
4. Address the missing values using an appropriate method
5. Run a check for duplicate values
6. Export the datasets


## Exercise 4.5

## Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## Import Data

In [2]:
#Create a path
path = r'C:\Users\dyl41\Documents\11-14-2022 Instacart Basket Analysis'

In [3]:
#Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [37]:
#Import orders.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [39]:
#Remove and reset index column
df_ords = df_ords.reset_index(drop = True)
df_ords = df_ords.drop(columns = 'Unnamed: 0')

## Data Consistency

## Mixed-Type Data

In [7]:
#Create a dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
#Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
#Convert 'mix' to string 
df_test['mix'] = df_test['mix'].astype('str')

## Finding Missing Values

In [13]:
#Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
#View the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
#Check new dataframe
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
#Describe dataframe
df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

In [18]:
#See shape of dataframe
df_prods.shape

(49693, 5)

In [21]:
# Define new dataset without the missing values in the products column -- the false
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
#See shape of new dataframe without missing values
df_prods_clean.shape

(49677, 5)

## Duplicates

In [23]:
#Finding duplicate values
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
#Check the duplicate values
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
#Check the shape of df_prods_clean
df_prods_clean.shape

(49677, 5)

In [29]:
#Drop Duplicates using new dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
#Check the shape of df_prods_clean_no_dups to see changes
df_prods_clean_no_dups.shape

(49672, 5)

In [31]:
#View the dataframe
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

## Export

In [34]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

## Question #1

In [45]:
#Use df.describe to check for consistency in df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

the max price in the prices column is 99,999; this warrants further investigation for possible error

In [47]:
df_prods[df_prods.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

Check revealed erroneous price for milk

## Question #2

In [49]:
#Check for mixed-type data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, 'mixed')

No mixed types found

## Question #4

In [48]:
#Run a check for missing values on df_ords dataframe
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The days_since_prior_order column has 206209 missing values. Using the .head function below,
it appears there is a pattern with the first order of each user_id displaying 'NaN'. Checking the max
user_id revealed that there are 206209 user id's; therefore, the NaN values correspond to the first order
of each user id.

In [53]:
df_ords.head(50)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   
11  2168274       2             1                   2                 11   
12  1501582       2             2                   5                 10   
13  1901567       2             3                   1                 10   
14   738281       2             4                   2                 10   
15  1673511       2             5                   3                 11   
16  1199898       2             6                   2                  9   
17  3194192       2             7                   2                 12   
18   788338       2             8                   1                 15   
19  1718559       2             9                   2                  9   
20  1447487       2            10                   1                 11   
21  1402090       2            11                   1                 10   
22  3186735       2            12                   1                  9   
23  3268552       2            13                   4                 11   
24   839880       2            14                   3                 10   
25  1492625       2            15                   1                 11   
26  1374495       3             1                   1                 14   
27   444309       3             2                   3                 19   
28  3002854       3             3                   3                 16   
29  2037211       3             4                   2                 18   
30  2710558       3             5                   0                 17   
31  1972919       3             6                   0                 16   
32  1839752       3             7                   0                 15   
33  3225766       3             8                   0                 17   
34  3160850       3             9                   0                 16   
35   676467       3            10                   3                 16   
36   521107       3            11                   0                 18   
37  1402502       3            12                   1                 15   
38  2774568       3            13                   5                 15   
39  3343014       4             1                   6                 11   
40  2030307       4             2                   4                 11   
41   691089       4             3                   4                 15   
42    94891       4             4                   5                 13   
43  2557754       4             5                   5                 13   
44   329954       4             6                   3                 12   
45  2717275       5             1                   3                 12   
46  1909121       5             2                   0                 16   
47  2267326       5             3                   3                 18   
48   157374       5             4                   1                 18   
49  2196797       5             5                   0                 11   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3         

In [61]:
df_ords.user_id.max()

206209

## Question #5

In [82]:
#Address the missing values using an appropriate method


In [76]:
#Create new dataframe from the old one
df_ords_new = df_ords

In [79]:
#Specify 'first_purchase' to be true if days_since_prior_order is NaN
df_ords_new['first_purchase'] = df_ords['days_since_prior_order'].isnull() == True

In [81]:
df_ords_new.head(50)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   
20   1447487        2            10                   1                 11   
21   1402090        2            11                   1                 10   
22   3186735        2            12                   1                  9   
23   3268552        2            13                   4                 11   
24    839880        2            14                   3                 10   
25   1492625        2            15                   1                 11   
26   1374495        3             1                   1                 14   
27    444309        3             2                   3                 19   
28   3002854        3             3                   3                 16   
29   2037211        3             4                   2                 18   
30   2710558        3             5                   0                 17   
31   1972919        3             6                   0                 16   
32   1839752        3             7                   0                 15   
33   3225766        3             8                   0                 17   
34   3160850        3             9                   0                 16   
35    676467        3            10                   3                 16   
36    521107        3            11                   0                 18   
37   1402502        3            12                   1                 15   
38   2774568        3            13                   5                 15   
39   3343014        4             1                   6                 11   
40   2030307        4             2                   4                 11   
41    691089        4             3                   4                 15   
42     94891        4             4                   5                 13   
43   2557754        4             5                   5                 13   
44    329954        4             6                   3                 12   
45   2717275        5             1                   3                 12   
46   1909121        5             2                   0                 16   
47   2267326        5             3                   3                 18   
48    157374        5             4                   1                 18   
49   2196797        5             5                   0                 11   

    days_since_prior_orde

Since imputation would likely skew the results and the other rows are still of value to the
dataframe and, therefore, should not be removed, I created a new column to specify that the
missing value was due to it being the customer's first purchase.

## Question #6

In [84]:
#Run a check for duplicate values
df_ords_dups = df_ords_new[df_ords_new.duplicated()]

In [92]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_purchase]
Index: []

No duplicates were found in the dataframe

## Question #7

In [93]:
#Export changes
df_ords_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

NameError: name 'df_ords' is not defined

In [3]:
import pandas as pd
import numpy as np
import os

In [4]:
path = r'C:\Users\dyl41\Documents\11-14-2022 Instacart Basket Analysis'

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
df_ords_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
 7   first_purchase          bool   
dtypes: bool(1), float64(1), int64(6)
memory usage: 186.0 MB


In [8]:
df_ords_new

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_purchase  
0                        8                     NaN            True  
1                        7                    15.0           False  
2                       12                    21.0           False  
3                        7                    29.0           False  
4                       15                    28.0           False  
...                    ...                     ...             ...  
3421078                 18                    29.0           False  
3421079                 10                    30.0           False  
3421080                 12                    18.0           False  
3421081                 12                     7.0           False  
3421082                 14                    30.0           False  

[3421083 rows x 8 columns]

In [9]:
#Check for mixed-type data in df_ords
for col in df_ords_new.columns.tolist():
  weird = (df_ords_new[[col]].applymap(type) != df_ords_new[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords_new[weird]) > 0:
    print (col, 'mixed')

In [10]:
df_ords_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
 7   first_purchase          bool   
dtypes: bool(1), float64(1), int64(6)
memory usage: 186.0 MB
